### ISSUE TITLE
[BUG] Wrong CSeq in ACK after Request Pending

### ISSUE THREAD

**Issue by AmduxGithub** [OpenSIPS/opensips#2236]
Hello fellow Voip enthusiasts,

we have encountered a bug in the current 3.0 Version regarding Cseq calculation after encountering a SIP 491 Request Pending on a a Re-Invite. 

**OpenSIPS version you are running**
version: opensips 3.0.2 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 9106ec5e5
main.c compiled on 22:20:13 Jun  6 2020 with gcc 7

We have not yet tried 3.1 as this is happening on a production machine cluster with thousends of concurrent calls but we did not see anything related in the patchnotes.
```
```
**Describe the bug**
_Enviremont_:
I do not think it is too relevant for this bug, but to describe the enviroment completely:

We run the opensips in question as a stateful proxy receiving  traffic to and from a customer on the one side, and interconnecting with Interconnection partners on the other side. 

The system is running with moderate load in AWS, with dedicated fibre towards the customer, and dedicated fibre towards interconnection partners, so there is a mix of public and private ip addresses and we do use ip alias and advertised_address. The setup is genereally working fine with millions of minutes switched per month.

_Actual Bugreport:_
Our customer receives an inbound call from an interconnection partner which is established and running.
After 15 minutes there is a session-refresh, where our customer sends two Re-Invites in very short time after another, these Re-Invites have different CSeqs, so the second one is not a retransmission. These two Re-Invites are a problem in the customers setup and another story, but it does triggers this bug reliable.

The interconnection partner sends a SIP 491 Request Pending Error for the second Cseq (the higher one) Re-Invite and sends a 200 ok for the first Cseq Re-Invite there after. 
Our customer then sends the ACK correctly for the first CSeq (the lower one) to our opensips proxy, which forwards the ACK but sets the second CSeq (the higher one) of the call that was rejected before from the interconnection partner.

The interconnection partner then sends retransmissions of the SIP 200 OK message as they did not receive an matching ACK and the call is finally hung up do to session timeout.

I have a ngrep trace attached as seperate TXT File.

**To Reproduce**
Send 2 Re-Invites in short time with increasing CSeq numbers, reject the high CSeq number and accept the lower CSeq Re-Invite. The state in opensips will then be wrong as it thinks the call is in the higher CSeq and the newly generated messages (ACK, BYE,..) will contain the wrong Cseq Value.

**Expected behavior**
The CSeq number should only be increased if the corresponding Invite was successful, if there is an error, the CSeq must be reverted to the value before the increase.

**Relevant System Logs**
I've attached a trace of an example call.
[cseq-problem.txt](https://github.com/OpenSIPS/opensips/files/5156217/cseq-problem.txt)

User phone numbers and domain names in from and to headers were replaced by 'a-number' and 'a-carrier.tld' respectively 'b-number' and 'b-carrier.tld' due to privacy concerns. 
The public ip-address 3.126.xxx.xxx and 3.127.xxx.xxx are another opensips instance under our control which act as gateways for the interconnection partner and can be regarded as such.

Following ip-addresses are used in the trace:

opensips ip-address: 172.31.2.20
Customer ip-address: 192.168.253.9
Interconnection Partner ip-address: 3.126.xxx.xxx

Packets of interest: 

Packets # 0:
INVITE from opensips to Customer 

Packet # 7: 
Re-Invite 1 from Customer to opensips due to Session Timer - Cseq: 1

Packet # 11: 
Re-Invite 2 from Customer to oepnsips due to Session Timer - CSeq: 2

Packet # 15: 
SIP 491 Request Pending from Interconnection Partner to opensips  - CSeq: 2

Packet # 19: 
200 OK from Interconnection Partner to opensips - CSeq: 1

Packet # 21: 
ACK from Customer to openips- Cseq: 1

Packet # 22: 
Ack from Opensips to Interconnection Partner - CSeq 2

**OS/environment information**
 - Operating System: Ubuntu 19.10

**Comment by razvancrainea** [comment]
Thank you for the report - this does seem to be a valid problem, but unfortunately doesn't seem to be fixable in an easy manner. We need to give it a thought for the proper solution. Will update the ticket as soon as we have one.

**Comment by wdavies-git** [comment]
This issue has been flagged high-priority for over a year, is anyone working on this? is fix forecast for any upcoming release? company I work with encountered this issue and had to put ugly config in to handle the glare events to avoid call failures.

**Comment by HichamINOX** [comment]
Hi,

this problem been solved in version 3 or is it still open ?
Any workaround ?

**Comment by ankogan** [comment]
Hi!

Have the same problem. When Re-Invite comes in microseconds after ACK to initial invite, the ACK packet sent to destication has CSeq of reinvite.

**Comment by HichamINOX** [comment]
Hi @ankogan ,

Here is what solved my problem, however the origin of the problem is not very clear, I cannot guarantee that it will work for you.

` if (has_totag()){
    #### Workaround pour les RE-invite rapides ####effet de retard introduit par t_newtran()? - Cseq issue git opensips 2236  ######
                if (is_method("INVITE"))
                {
                        t_newtran();
                }
   }`

**Comment by fbinet** [comment]
> Hi @ankogan ,
> 
> Here is what solved my problem, however the origin of the problem is not very clear, I cannot guarantee that it will work for you.
> 
> ` if (has_totag()){ #### Workaround pour les RE-invite rapides ####effet de retard introduit par t_newtran()? - Cseq issue git opensips 2236 ###### if (is_method("INVITE")) { t_newtran(); } }`

Indeed a strange issue. I have the same thing happening on very-quick-reinvites, tried out that t_newtran this morning, waiting for customer results but I still can see their switch replying "Invalid CSeq" so my hope are not very high.

Interesting that there is not more reports than this, but I think the call drop might happen only when connected to some type of equipment. the customer having most issues with this is using asterisk 16.2 with PJSIP

**Comment by razvancrainea** [comment]
This ticket seems to be a duplicate of #1071, for which a fix has been proposed and pushed upstream on all supported branches.
Marking this ticket as closed as it is outdated anyway (3.0 is no longer supported) - please follow up on #1071 with any updates.
